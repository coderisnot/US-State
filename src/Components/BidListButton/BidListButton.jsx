import React, { Component } from 'react';
import PropTypes from 'prop-types';
import FontAwesome from 'react-fontawesome';
import { existsInNestedObject } from '../../utilities';
import { BID_RESULTS } from '../../Constants/PropTypes';

class BidListButton extends Component {

  constructor(props) {
    super(props);
    this.toggleSaved = this.toggleSaved.bind(this);
  }

  getIsSaved() {
    // Is the id in the array? If so, return true
    const { compareArray, id } = this.props;
    return existsInNestedObject(id, compareArray);
  }

  toggleSaved() {
    const { toggleBidPosition, id } = this.props;
    // pass the id and the "remove" param
    toggleBidPosition(id, this.getIsSaved());
  }

  render() {
    // is the bid currently saved?
    // save value and avoid interrogating the array more than once
    const bidIsSaved = this.getIsSaved();
    const text = bidIsSaved ? 'Remove from Bid List' : 'Add to Bid List';
    const iconClass = bidIsSaved ? 'minus-circle' : 'plus-circle';
    const style = {
      pointerEvents: this.props.isLoading ? 'none' : 'inherit',
    };
    const { className } = this.props;
    return (
      <button className={className} style={style} onClick={this.toggleSaved}>
        <span className="button-icon">
          <FontAwesome name={iconClass} />
        </span>
        <span>{text}</span>
      </button>
    );
  }
}

BidListButton.propTypes = {
  id: PropTypes.number.isRequired,
  toggleBidPosition: PropTypes.func.isRequired,
  compareArray: BID_RESULTS,
  isLoading: PropTypes.bool,
  className: PropTypes.string,
};

BidListButton.defaultProps = {
  compareArray: [],
  isLoading: false,
  className: 'bid-list-button',
};

export default BidListButton;
