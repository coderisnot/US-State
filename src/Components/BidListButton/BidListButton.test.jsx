import { shallow } from 'enzyme';
import sinon from 'sinon';
import React from 'react';
import toJSON from 'enzyme-to-json';
import BidListButton from './BidListButton';

describe('BidListButtonComponent', () => {
  const bidList = [{ position: { id: 1 } }];
  const bidListFalse = [{ position: { id: 2 } }];

  it('is defined', () => {
    const wrapper = shallow(
      <BidListButton
        id={1}
        toggleBidPosition={() => {}}
        compareArray={bidList}
        isLoading={false}
      />,
    );
    expect(wrapper).toBeDefined();
  });

  it('handles text change when savedState is true', () => {
    const wrapper = shallow(
      <BidListButton
        id={1}
        toggleBidPosition={() => {}}
        compareArray={bidList}
        isLoading={false}
      />,
    );
    expect(wrapper.find('Add')).toBeDefined();
  });

  it('handles text change when savedState is false', () => {
    const wrapper = shallow(
      <BidListButton
        id={1}
        toggleBidPosition={() => {}}
        compareArray={bidListFalse}
        isLoading={false}
      />,
    );
    expect(wrapper.find('Remove')).toBeDefined();
  });

  it('can call the toggleSaved function', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <BidListButton
        id={1}
        toggleBidPosition={spy}
        compareArray={bidListFalse}
        isLoading={false}
      />,
    );
    wrapper.find('button').simulate('click');
    sinon.assert.calledOnce(spy);
  });

  it('matches snapshot when the user can add the position', () => {
    const wrapper = shallow(
      <BidListButton
        id={1}
        toggleBidPosition={() => {}}
        compareArray={bidListFalse}
        isLoading={false}
      />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('matches snapshot when the user can remove the position', () => {
    const wrapper = shallow(
      <BidListButton
        id={1}
        toggleBidPosition={() => {}}
        compareArray={bidList}
        isLoading={false}
      />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
