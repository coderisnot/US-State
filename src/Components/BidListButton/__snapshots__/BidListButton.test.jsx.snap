// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BidListButtonComponent matches snapshot when the user can add the position 1`] = `
<button
  className="bid-list-button"
  onClick={[Function]}
  style={
    Object {
      "pointerEvents": "inherit",
    }
  }
>
  <span
    className="button-icon"
  >
    <FontAwesome
      name="plus-circle"
    />
  </span>
  <span>
    Add to Bid List
  </span>
</button>
`;

exports[`BidListButtonComponent matches snapshot when the user can remove the position 1`] = `
<button
  className="bid-list-button"
  onClick={[Function]}
  style={
    Object {
      "pointerEvents": "inherit",
    }
  }
>
  <span
    className="button-icon"
  >
    <FontAwesome
      name="minus-circle"
    />
  </span>
  <span>
    Remove from Bid List
  </span>
</button>
`;
